### PR TITLE
Add support for polygons

### PR DESCRIPTION
**Pull request by acq**
Add the possibility to draw polygons on an AirMapView (on top of markers, polylines and circles)

### CODE REVIEW

File: library/src/main/java/com/airbnb/android/airmapview/AirMapView.java [hunk 1]
```diff
@@ -279,6 +279,22 @@ public boolean removePolyline(AirMapPolyline polyline) {
     return false;
   }
 
+  public boolean addPolygon(AirMapPolygon polygon) {
```
Comment (felipecsl) [inline library/src/main/java/com/airbnb/android/airmapview/AirMapView.java]: can you use `AirMapPolygon<?>` instead?

File: library/src/main/java/com/airbnb/android/airmapview/AirMapView.java [hunk 2]
```diff
@@ -279,6 +279,22 @@ public boolean removePolyline(AirMapPolyline polyline) {
...
+    if (isInitialized()) {
+      mapInterface.addPolygon(polygon);
+      return true;
+    }
+    return false;
+  }
+
+  public boolean removePolygon(AirMapPolygon polygon) {
```
Comment (felipecsl) [inline library/src/main/java/com/airbnb/android/airmapview/AirMapView.java]: same here

File: library/src/main/java/com/airbnb/android/airmapview/WebViewMapFragment.java [hunk 3]
```diff
@@ -236,6 +236,26 @@ public void setOnMarkerClickListener(OnMapMarkerClickListener listener) {
...
+    try {
+      JSONArray array = new JSONArray();
+      for (LatLng point : polygon.getPolygonOptions().getPoints()) {
+        JSONObject json = new JSONObject();
+        json.put("lat", point.latitude);
+        json.put("lng", point.longitude);
+        array.put(json);
+      }
+
+      webView.loadUrl(String.format("javascript:addPolygon(" + array.toString() + ", %1$d, %2$d, %3$d, %4$d);", polygon.getId(), (int) polygon.getPolygonOptions().getStrokeWidth(), polygon.getPolygonOptions().getStrokeColor(), polygon.getPolygonOptions().getFillColor()));
```
Comment (felipecsl) [inline library/src/main/java/com/airbnb/android/airmapview/WebViewMapFragment.java]: can you wrap this line in multiple, please? It's too long. 
Also use `Locale.US` as the first argument to `String.format`

**Comment by felipecsl** [comment]
Looking pretty good overall. Can you add a screenshot please?

**Comment by acq** [comment]
Thanks for the feedback.
I just used <T> instead of <?>, if that's alright

Screenshots have also been updated

File: library/src/main/assets/google_map.html [hunk 4]
```diff
@@ -182,6 +182,35 @@
   polylines = {};
 }
 
+function addPolygon(points, id, strokeWidth, strokeColor, fillColor) {
+  var pointsArray = [];
+
+  for (var i = 0; i < points.length; i++) {
+    var point = points[i];
+    var position = new google.maps.LatLng(point["lat"], point["lng"]);
```
Comment (felipecsl) [inline library/src/main/assets/google_map.html]: it seems to me like you're not using this variable `position` anywhere?

File: library/src/main/java/com/airbnb/android/airmapview/AirMapInterface.java [hunk 5]
```diff
@@ -194,4 +195,16 @@
 
   /** Sets the type of map tiles that should be displayed */
   void setMapType(MapType type);
-}
\ No newline at end of file
+
+  /**
+   * Construct a polygon with the given {@link LatLng} points
+   */
+  <T> void addPolygon(AirMapPolygon<T> polygon);
```
Comment (felipecsl) [inline library/src/main/java/com/airbnb/android/airmapview/AirMapInterface.java]: the javadocs comment seems outdated. You refer to a LatLng but it takes an AirMapPolygon

File: library/src/main/java/com/airbnb/android/airmapview/AirMapPolygon.java [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+
+public class AirMapPolygon<T> {
+
+  private static final int STROKE_WIDTH = 1;
+  private static final int STROKE_COLOR = Color.BLUE;
+
+  private T              object;
+  private long           id;
+  private PolygonOptions polygonOptions;
+  private Polygon        googlePolygon;
```
Comment (felipecsl) [inline library/src/main/java/com/airbnb/android/airmapview/AirMapPolygon.java]: just use a single space for indentation
also final the id and object at least

**Comment by acq** [comment]
Actually, Polyline was suffering from the same issues (Javadoc outdated, unused local variable), so I updated the appropriate sections

File: library/src/main/assets/google_map.html [hunk 7]
```diff
@@ -182,6 +181,34 @@
...
+  for (var i = 0; i < points.length; i++) {
+    var point = points[i];
+    pointsArray.push(point);
+  }
+
+  strokeColor = formatColor(strokeColor);
+  fillColor = formatColor(fillColor);
+  var polygon = new google.maps.Polygon({
+    map: map,
+    paths: pointsArray,
```
Comment (felipecsl) [inline library/src/main/assets/google_map.html]: can you just use `points` here directly? it seems like `pointsArray` is identical

File: library/src/main/assets/google_map.html [hunk 8]
```diff
@@ -182,6 +181,34 @@
   polylines = {};
 }
 
+function addPolygon(points, id, strokeWidth, strokeColor, fillColor) {
+  var pointsArray = [];
+
+  for (var i = 0; i < points.length; i++) {
```
Comment (ljharb) [inline library/src/main/assets/google_map.html]: why is this not `var pointsArray = points.slice();` if you want a copy?

File: library/src/main/assets/mapbox.html [hunk 9]
```diff
@@ -143,6 +142,28 @@
   polylines = {};
 }
 
+function addPolygon(points, id, width, strokeColor, fillColor) {
+  var pointsArray = [];
+
+  for (var i = 0; i < points.length; i++) {
+    var point = points[i];
+    pointsArray.push(point);
```
Comment (felipecsl) [inline library/src/main/assets/mapbox.html]: same here

File: library/src/main/java/com/airbnb/android/airmapview/AirMapPolygon.java [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+    return googlePolygon;
+  }
+
+  public void setGooglePolygon(Polygon googlePolygon) {
+    this.googlePolygon = googlePolygon;
+  }
+
+  public static class Builder<T> {
+    private final PolygonOptions polygonOptions = new PolygonOptions();
+    private T    object;
```
Comment (felipecsl) [inline library/src/main/java/com/airbnb/android/airmapview/AirMapPolygon.java]: one space between T and object

**Comment by acq** [comment]
Looks like there was indeed no need for pointsArray.
Should I squash these commits in one?

**Comment by felipecsl** [comment]
looks good to me. @acq yes please squash them?
@petzel or @nwadams wanna take a quick look?

File: library/src/main/assets/mapbox.html [hunk 11]
```diff
@@ -115,16 +115,8 @@
...
 
 function addPolyline(points, id, width, color) {
-  var pointsArray = [];
-
-  for (var i = 0; i < points.length; i++) {
-    var point = points[i];
-    var position = L.latLng(point["lat"], point["lng"]);
-    pointsArray.push(point);
-  }
-
```
Comment (petzel) [inline library/src/main/assets/mapbox.html]: :+1: nice, thanks for cleaning this up

File: library/src/main/java/com/airbnb/android/airmapview/AirMapInterface.java [hunk 12]
```diff
@@ -194,4 +197,18 @@
...
+   * Add the given polygon to the map
+   *
+   * @param polygon {@link AirMapPolygon} instance to add
+   */
+  <T> void addPolygon(AirMapPolygon<T> polygon);
+
+  /**
+   * Remove the given {@link Polygon}
+   *
+   * @param polygon the {@link Polygon} to remove
```
Comment (petzel) [inline library/src/main/java/com/airbnb/android/airmapview/AirMapInterface.java]: the `link` here is to a `Polygon`, not an `AirMapPolygon` (same with line 209)

**Comment by petzel** [comment]
Looks great! Please do squash before merging. 

:ship:

**Comment by acq** [comment]
Javadoc links have been fixed.
Commits have been squashed.
Resulting commit has been rebased on latest master.

:ship: ?

**Pull request merged**